agera/src/main/java/com/google/android/agera/Observables.java

```diff
@@ -138,31 +139,75 @@ public static UpdateDispatcher updateDispatcher(
     return new AsyncUpdateDispatcher(activationHandler);
   }
 
-  private static final class CompositeObservable extends BaseObservable implements Updatable {
+  private static final class CompositeObservable implements Observable {
+    @NonNull
+    private static final Updatable[] NO_UPDATABLES = new Updatable[0];
     @NonNull
     private final Observable[] observables;
+    @NonNull
+    private Updatable[] updatables;
 
     CompositeObservable(@NonNull final Observable... observables) {
       this.observables = observables;
+      this.updatables = NO_UPDATABLES;
     }
 
     @Override
-    protected void observableActivated() {
+    public synchronized void addUpdatable(@NonNull final Updatable updatable) {
+      boolean added = false;
+      Updatable localUpdatable = null;
+      for (int index = 0; index < updatables.length; index += 2) {
+        if (updatables[index] == updatable) {
+          throw new IllegalStateException("Updatable already added, cannot add.");
+        }
+        if (updatables[index] == null) {
+          updatables[index] = updatable;
+          localUpdatable = new WrapperUpdatable(updatable);
+          updatables[index + 1] = localUpdatable;
+          added = true;
+          break;
```

Comment:
same goes here.

you might instead want to save the `int indexToAdd` (init to -1) instead of `boolean added`, so you do one thorough check before acting, throwing exception if necessary, and then do the "if not -1 then add to slot; otherwise expand array and add to end" afterwards.